sympy/matrices/matrices.py

```diff
@@ -1400,7 +1577,44 @@ def denom_clean(l):
             ret = [(val.evalf(chop=chop), mult, [v.evalf(chop=chop) for v in es]) for val, mult, es in ret]
         return ret
 
-    def is_diagonalizable(self, reals_only=False, **kwargs):
+    def is_diagonalizable_with_eigen(self, reals_only=False, dotprodsimp=None, **kwargs):
+        """See is_diagonalizable. This function returns the bool along with the
+        eigenvectors to avoid calculating them again in functions like
+        ``diagonalize``."""
+
+        if 'clear_cache' in kwargs:
+            SymPyDeprecationWarning(
+                feature='clear_cache',
+                deprecated_since_version=1.4,
+                issue=15887
+            ).warn()
+        if 'clear_subproducts' in kwargs:
+            SymPyDeprecationWarning(
+                feature='clear_subproducts',
+                deprecated_since_version=1.4,
+                issue=15887
+            ).warn()
```

Comment:
What about moving the deprecation warnings to `is_diagonalizable`, and also delete arbitrary `kwargs` here?

Reply:
Done...